deepgram/errors.py

```diff
@@ -0,0 +1,87 @@
+from typing import Union, Optional
+import websockets.exceptions
+import urllib.error
+import aiohttp
+import uuid
+
+
+class DeepgramError(Exception):
+    pass
+
+
+class DeepgramSetupError(DeepgramError, ValueError):
+    pass
+
+
+class DeepgramApiError(DeepgramError):
+    """An error returned by the Deepgram API.
+
+    If the error was raised by an http client, the client's error message
+    is accessible via the `http_library_error` field. This may be useful
+    for handling different error codes, such as 429s or 503s.
+
+    The `error` field is set to the API's error message (dict), if avilable.
+    Otherwise the `error` field is set to the parent exception's message (str).
+    
+    The `warning` field is set to the API's warning messages (list[str]), if available.
+
+    The `request_id` field is set to the API's request ID, if available.
+    """
+    def __init__(
+        self,
+        *args: object,
+        http_library_error: Optional[
+            Union[
+                urllib.error.HTTPError,
+                urllib.error.URLError,
+                websockets.exceptions.InvalidHandshake,
+                aiohttp.ClientResponseError,
+                aiohttp.ClientError,
+            ]
+        ] = None,
+    ):
+        super().__init__(*args)
+        self.http_library_error = http_library_error
+        self.error: Union[str, dict]  # If you change the type, change it in the docstring as well!
+        self.warnings: Optional[list[str]] = None  # If you change the type, change it in the docstring as well!
+        self.request_id: Optional[uuid.UUID] = None
+        self.http_error_status: Optional[int] = None
+
+        # Set the `error`, `warning`, and `request_id` fields
+        if isinstance(args[0], dict) and "err_msg" in args[0]:
+            self.error = args[0]["err_msg"]
+            if "metadata" in args[0] and "warnings" in args[0]["metadata"]:
+                self.warnings = args[0]["metadata"]["warnings"]
+            elif "warnings" in args[0]:  # Occurs when `raise_warnings_as_errors` is enabled
+                self.warnings = args[0]["warnings"]
+            if "metadata" in args[0] and "request_id" in args[0]["metadata"]:  # Occurs when Deepgram returns a success response (for warnings)
+                self.request_id = uuid.UUID(args[0]["request_id"])
+            elif "request_id" in args[0]:  # Occurs when Deepgram returns a failed response
+                self.request_id = uuid.UUID(args[0]["request_id"])
+        elif isinstance(args[0], str):
+            self.error = args[0]
+        else:
+            self.error = str(args[0])
```

Comment:
You're referring to `args[0]` a lot here, it would be helpful for readability to give that a named variable so it's clearer what that arg is.

Reply:
Thanks, fixed. I created a new variable: `error_or_warning_data = args[0]`